simulator/infra/replacement/t/unit_test.cpp

```diff
@@ -36,34 +36,79 @@ TEST_CASE( "Check_set_to_erase_method")
     CHECK( test_lru_module->update() == 2);
 }
 
-TEST_CASE( "Check_Untouched_way")
+
+TEST_CASE( "LRU: Check_Untouched_way")
 {
     auto test_lru_module = create_cache_replacement( "LRU", 3);
     test_lru_module->touch( 0);
     test_lru_module->touch( 2);
     CHECK( test_lru_module->update() == 1);
 }
 
-
-TEST_CASE( "Check_allocate_method")
+TEST_CASE( "LRU: Check_allocate_method")
 {
     auto test_lru_module = create_cache_replacement( "LRU", 3);
     CHECK_NOTHROW( test_lru_module->allocate( 0));
 }
 
-TEST_CASE( "Check_erase_lru_element_method")
+TEST_CASE( "LRU: Check_erase_lru_element_method")
 {
     auto test_lru_module = create_cache_replacement( "LRU", 3);
     CHECK_NOTHROW(test_lru_module->allocate(1));
 }
 
-TEST_CASE( "Check_get_ways_method")
+
+TEST_CASE( "LRU && Pseudo-LRU: Check_get_ways_method")
 {
     auto test_lru_module = create_cache_replacement( "LRU", 3);
+    auto test_pseudo_lru_module = create_cache_replacement( "Pseudo-LRU", 3);
     CHECK(test_lru_module->get_ways() == 3);
+    CHECK(test_pseudo_lru_module->get_ways() == 3);
+}
+
+TEST_CASE( "LRU: Check_bad_string_pass_to_factory_method")
+{
+    CHECK_THROWS_AS( create_cache_replacement( "BAD STRING", 3), CacheReplacementException);
+}
+
+TEST_CASE( "Pseudo-LRU: Check_forbidden_to_call_methods")
+{
+    auto test_pseudo_lru_module = create_cache_replacement( "Pseudo-LRU", 3);
```

Comment:
How 3-way pseudo LRU is supposed to work with binary tree?

Reply:
Parent, that contains empty node or branch, always points at not empty one. This works bad, as for example 257 way-cache will treat the last way as the LRU one in two cases. There may help some trick with rebalancing the tree, but anyway it won't work good with number of ways far from the powers of 2. Does it really have to?

Reply:
Yes, non power-2 pseudo LRUs are tricky. Let's stick to power-2 only modes, and throw an exception if someone tries to use something different.